src/iop/filmicrgb.c

```diff
@@ -1502,6 +1504,363 @@ static inline void filmic_chroma_v2_v3(const float *const restrict in, float *co
 }
 
 
+#ifdef _OPENMP
+#pragma omp declare simd uniform(matrix) aligned(in, out:16) aligned(matrix:64)
+#endif
+static inline void pipe_RGB_to_Ych(const dt_aligned_pixel_t in, const dt_colormatrix_t matrix, dt_aligned_pixel_t out)
+{
+  dt_aligned_pixel_t LMS = { 0.f };
+  dt_aligned_pixel_t Yrg = { 0.f };
+
+  // go from pipeline RGB to CIE 2006 LMS D65
+  dot_product(in, matrix, LMS);
+
+  // go from CIE LMS 2006 to Kirk/Filmlight Yrg
+  LMS_to_Yrg(LMS, Yrg);
+
+  // rewrite in polar coordinates
+  Yrg_to_Ych(Yrg, out);
+}
+
+
+#ifdef _OPENMP
+#pragma omp declare simd uniform(matrix) aligned(in, out:16) aligned(matrix:64)
+#endif
+static inline void Ych_to_pipe_RGB(const dt_aligned_pixel_t in, const dt_colormatrix_t matrix, dt_aligned_pixel_t out)
+{
+  dt_aligned_pixel_t LMS = { 0.f };
+  dt_aligned_pixel_t Yrg = { 0.f };
+
+  // rewrite in cartesian coordinates
+  Ych_to_Yrg(in, Yrg);
+
+  // go from Kirk/Filmlight Yrg to CIE LMS 2006
+  Yrg_to_LMS(Yrg, LMS);
+
+  // go from CIE LMS 2006 to pipeline RGB
+  dot_product(LMS, matrix, out);
+
+  for_each_channel(c, aligned(out)) out[c] = fmaxf(out[c], 0.f);
+}
+
+static inline void filmic_desaturate_v4(const dt_aligned_pixel_t Ych_original, dt_aligned_pixel_t Ych_final, const float saturation)
+{
+  // Note : Ych is normalized trough the LMS conversion,
+  // meaning c is actually a saturation (saturation ~= chroma / brightness).
+  // So copy-pasting c and h from a different Y is equivalent to
+  // tonemapping with a norm, which is equivalent to doing exposure compensation :
+  // it's saturation-invariant, aka chroma will get increased
+  // if Y is increased, and the other way around.
+  const float chroma_original = Ych_original[1] * Ych_original[0];  // c2
+  float chroma_final = Ych_final[1] * Ych_final[0];                 // c1
+
+  // fit a linear model `chroma = f(y)`:
+  // `chroma = c1 + (yc - y1) * (c2 - c1) / (y2 - y1)`
+  // where `(yc - y1)` is user-defined as `saturation * (y2 - y1)`
+  // so `chroma = c1 + saturation * (c2 - c1)`
+  // when saturation = 0, we stay at the saturation-invariant final chroma
+  // when saturation > 1, we go back towards the initial chroma before tone-mapping
+  // when saturation < 0, we amplify the initial -> final chroma change
+  const float delta_chroma = saturation * (chroma_original - chroma_final);
+
+  const int filmic_brightens = (Ych_final[0] > Ych_original[0]);
+  const int filmic_resat = (chroma_original < chroma_final);
+  const int filmic_desat = (chroma_original > chroma_final);
+  const int user_resat = (saturation > 0.f);
+  const int user_desat = (saturation < 0.f);
+
+  chroma_final = (filmic_brightens && filmic_resat)
+                      ? chroma_original // force original lower sat if brightening
```

Comment:
This may be a personal opinion, but this seems to produce a bit of a washed-out look in the highlights. Personally I would rather preserve the saturation even in the highlights. After all, there's `color balance rgb` which can be used to achieve that saturation roll-off in the highlights, but if filmic does this by default, it's harder to "undo" in a `color balance rgb` instance which usually comes before filmic in the pixel pipe.

Reply:
Ah, I see what you meant in the comment below. v5 has an explicit desaturation curve to white. v6 doesn't. This was intended to provide a better intermediate between no desaturation at all and the full-blown desaturation of the split RGB method. Maybe we could set that to the average of `chroma_original` and `chroma_final` ?

Reply:
> Maybe we could set that to the average of chroma_original and chroma_final ?

That could probably be a good enough middle ground. However, I would still consider the option of leaving the desaturation out entirely for the chrominance preserving modes - one could always desaturate the highlights earlier in the pipe in e.g. `color balance rgb`. The "add basic colorfulness" preset already does this.

Reply:
I personally find that leaving chroma as-is yields oversaturated colors. Keep in mind that people come from the split-RGB world where highlights are indeed heavily desaturated (which is also consistent with film sensitometry). I would make the reverse case where you can always resaturate highlights through color balance, but that average seems a better default starting point in general.